examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/DiscoveryExampleFragment.java

```diff
@@ -0,0 +1,390 @@
+import android.widget.TextView;
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import com.R;
+import com.chip.casting.SuccessCallback;
+import com.matter.casting.core.CastingPlayer;
+import com.matter.casting.core.CastingPlayerDiscovery;
+import com.matter.casting.core.MatterCastingPlayerDiscovery;
+import com.matter.casting.support.MatterError;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class DiscoveryExampleFragment extends Fragment {
+  private static final String TAG = DiscoveryExampleFragment.class.getSimpleName();
+  private boolean discovering = false;
+  private TextView matterDiscoveryMessageTextView;
+  private int discoveryAttempt = 0;
+  private static final ScheduledExecutorService executorService =
+      Executors.newSingleThreadScheduledExecutor();
+  private ScheduledFuture scheduledFutureTask;
+  private static final List<CastingPlayer> castingPlayerList = new ArrayList<>();
+  // private FailureCallback failureCallback;
+  private static SuccessCallback<CastingPlayer> discoverySuccessCallback;
+
+  // Get a singleton instance of the MatterCastingPlayerDiscovery
+  private static final CastingPlayerDiscovery matterCastingPlayerDiscovery =
+      MatterCastingPlayerDiscovery.getInstance();
+
+  /**
+   * Implementation of a CastingPlayerChangeListener used to listen to changes in the discovered
+   * CastingPlayers
+   */
+  private static final CastingPlayerDiscovery.CastingPlayerChangeListener
+      castingPlayerChangeListener =
+          new CastingPlayerDiscovery.CastingPlayerChangeListener() {
+            private final String TAG =
+                CastingPlayerDiscovery.CastingPlayerChangeListener.class.getSimpleName();
+
+            @Override
+            public void onAdded(CastingPlayer castingPlayer) {
+              if (castingPlayer != null) {
+                Log.i(
+                    TAG,
+                    "onAdded() Discovered CastingPlayer deviceId: " + castingPlayer.getDeviceId());
+                // Display CastingPlayer info on the screen
+                if (discoverySuccessCallback != null) {
+                  discoverySuccessCallback.handle(castingPlayer);
+                } else {
+                  Log.e(
+                      TAG,
+                      "onAdded() Warning: DiscoveryExampleFragment UX discoverySuccessCallback not set");
+                }
+              } else {
+                Log.d(TAG, "onAdded()");
+                // Attempt to invoke interface method on a null object reference will throw an error
+                Log.d(
+                    TAG,
+                    "onAdded() Discovered CastingPlayer with deviceId: "
+                        + castingPlayer.getDeviceId());
+              }
```

Comment:
discoverySuccessCallback seems to be adding another level of indirection which does not add much value in this case. I think the code would be more readable if you inlined the logic from the discoverySuccessCallback here.